### ISSUE TITLE
Rewrite signal handling in SGX

### ISSUE THREAD

**Issue by chiache** [oscarlab/graphene#292]
The same design change is made for Linux (#281). The SGX PAL is left for future work but should be addressed in near future.

**Comment by dimakuv** [comment]
Unfortunately, the solution in #2533 seems to introduce significant performance overheads (2-7% on some tested workloads). This doesn't sound acceptable, so we have to come up with another solution. This means that we'll have to postpone solving this issue to a later time.

**Comment by boryspoplawski** [comment]
1. It's #2649 not #2533
2. Why is that unacceptable? Most tested workloads exhibited <=2% run time degradation. With the improved code manageability and bug fixes it might be worth merging. Or might not. I would say it needs consideration, but I would definitely not dismiss it based on 1 workload with 7% degradation.

**Comment by mkow** [comment]
Yup, only one workload had this bad 7% degradation and everything else had negligible slowdowns. Before dismissing this solution we should check this one workload more closely (this was TF ResNet50, if I remember correctly?), maybe it's doing something really weird.
